lib/HTFeed/BackupExpiration.pm

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/perl
+package HTFeed::BackupExpiration;
+
+use strict;
+use warnings;
+use HTFeed::Config qw(get_config);
+use HTFeed::DBTools qw(get_dbh);
+use HTFeed::Volume;
+use Carp;
+use Log::Log4perl qw(get_logger);
+use File::Temp;
+
+sub new {
+  my $class = shift;
+
+  my $self = {
+    storage_name => undef,
+    @_
+  };
+
+  unless ($self->{storage_name}) {
+    croak "$class cannot be constructed without a storage name";
+  }
+
+  bless($self, $class);
+  return $self;
+}
+
+sub run {
+  my $self = shift;
+
+  my $config = get_config('storage_classes');
+  my $storage_config = $config->{$self->{storage_name}};
+  # Get a list of all volumes with this storage name that have one or more old
+  # versions and one or more new versions. This query is about finding sets of
+  # volumes, nothing more granular.
+  my $sth = get_dbh()->prepare(<<'SQL');
+    SELECT namespace,id,MIN(version) AS min_version,MAX(version) AS max_version
+    FROM feed_backups
+    WHERE deleted IS NULL
+      AND storage_name=?
+    GROUP BY namespace,id
+    HAVING COUNT(*) > 1
+      AND min_version < DATE_FORMAT(DATE_SUB(NOW(), INTERVAL 180 DAY),"%Y%m%d%H%i%S")
+      AND max_version > DATE_FORMAT(DATE_SUB(NOW(), INTERVAL 180 DAY),"%Y%m%d%H%i%S")
+SQL
+
+  $sth->execute($self->{storage_name});
+  while (my $row = $sth->fetchrow_hashref) {
+    # Get the versions in this group that can be deleted.
+    # The previous query has verified that there is >= 1 unexpired version.
+    my $versions_sth = get_dbh()->prepare(<<'SQL');
+      SELECT version
+      FROM feed_backups
+      WHERE deleted IS NULL
+        AND storage_name=?
+        AND namespace=?
+        AND id=?
+        AND version < DATE_FORMAT(DATE_SUB(NOW(), INTERVAL 180 DAY),"%Y%m%d%H%i%S")
```

Comment:
This will need adjustment given the previous comment -- I'm not sure what the easiest way to express that is, but I think we want to:

- in no case should we delete the most recent version (even if it is more than 6 months old)
- in no case should we delete anything newer than 6 months old